### PR TITLE
bpo-32253: Deprecate with statement and bare await for asyncio locks

### PR DESCRIPTION
**Pull request by asvetlov**
https://bugs.python.org/issue32253

### CODE REVIEW

File: Doc/library/asyncio-sync.rst [hunk 1]
```diff
@@ -51,38 +50,12 @@ Lock
...
-   :meth:`acquire` is a coroutine and should be called with ``yield from``.
+   :meth:`acquire` is a coroutine and should be called with ``await``.
 
-   Locks also support the context management protocol.  ``(yield from lock)``
-   should be used as the context manager expression.
+   Locks also support the :ref:`context management protocol <async-with-locks>`.
 
    This class is :ref:`not thread safe <asyncio-multithreading>`.
 
-   Usage::
```
Comment (1st1) [inline Doc/library/asyncio-sync.rst]: ~Let's have one usage example with 'async with ...' syntax.~

Comment (1st1) [reply Doc/library/asyncio-sync.rst]: Never mind, I see that you added a new section for that...

File: Doc/library/asyncio-sync.rst [hunk 2]
```diff
@@ -285,11 +259,54 @@ Semaphore
...
+   This class is :ref:`not thread safe <asyncio-multithreading>`.
+
+
+.. _async-with-locks:
+
+Using locks, conditions, and semaphores in the :keyword:`async with` statement
+------------------------------------------------------------------------------
+
+All of the objects provided by this module that have :meth:`acquire`
+and :meth:`release` methods can be used as context managers for a
```
Comment (1st1) [inline Doc/library/asyncio-sync.rst]: "can be used as context managers for a" -> "can be used in "async with" statements"

File: Doc/library/asyncio-sync.rst [hunk 3]
```diff
@@ -285,11 +259,54 @@ Semaphore
...
+
+   await lock.acquire()
+   try:
+       # do something...
+   finally:
+       lock.release()
+
+Currently, :class:`Lock`, :class:`Condition`,
+:class:`Semaphore`, and :class:`BoundedSemaphore` objects may be used as
+:keyword:`async with` statement context managers.
```
Comment (1st1) [inline Doc/library/asyncio-sync.rst]: This is a bit repetitive... I'd merge this paragraph with the first one:

":class:`Lock`, :class:`Condition`, :class:`Semaphore`, and :class:`BoundedSemaphore` can be used in "async with" statements."

File: Doc/library/asyncio-sync.rst [hunk 4]
```diff
@@ -285,11 +259,54 @@ Semaphore
...
+
+``async with lock`` is the preferable way for locking asyncio
+synchronization primitives.
+
+Explicit :meth:`acquire` / :meth:`release` calls should be used if
+locking/unlocking is split into different functions.
```
Comment (1st1) [inline Doc/library/asyncio-sync.rst]: I don't think we need these two paragraphs.  People usually naturally prefer 'with' blocks when they are available.

File: Doc/library/asyncio-sync.rst [hunk 5]
```diff
@@ -285,11 +259,54 @@ Semaphore
...
+
+.. deprecated:: 3.7
+
+   Lock acquiring on ``await lock`` or ``yield from lock`` and
```
Comment (1st1) [inline Doc/library/asyncio-sync.rst]: "Lock acquiring on" -> "Lock acquiring using"

**Comment by asvetlov** [comment]
All notes are fixed

File: Doc/library/asyncio-sync.rst [hunk 6]
```diff
@@ -260,6 +232,8 @@ Semaphore
    defaults to ``1``. If the value given is less than ``0``, :exc:`ValueError`
    is raised.
 
+   Locks also support the :ref:`context management protocol <async-with-locks>`.
```
Comment (1st1) [inline Doc/library/asyncio-sync.rst]: Semaphores also support...

File: Doc/library/asyncio-sync.rst [hunk 7]
```diff
@@ -285,11 +259,46 @@ Semaphore
...
+----------------
 
 .. class:: BoundedSemaphore(value=1, \*, loop=None)
 
    A bounded semaphore implementation. Inherit from :class:`Semaphore`.
 
    This raises :exc:`ValueError` in :meth:`~Semaphore.release` if it would
    increase the value above the initial value.
+
+   Locks also support the :ref:`context management protocol <async-with-locks>`.
```
Comment (1st1) [inline Doc/library/asyncio-sync.rst]: BoundedSemaphore also support...

**Comment by 1st1** [comment]
Spotted a few more nits; other than that it looks good

File: Doc/library/asyncio-sync.rst (hunk 6 above)
Comment (asvetlov) [reply Doc/library/asyncio-sync.rst]: Nice catch

File: Doc/library/asyncio-sync.rst (hunk 7 above)
Comment (asvetlov) [reply Doc/library/asyncio-sync.rst]: Done

**Comment by asvetlov** [comment]
Fixed

File: Doc/library/asyncio-sync.rst [hunk 8]
```diff
@@ -139,7 +139,8 @@ Condition
    object, and it is used as the underlying lock.  Otherwise,
    a new :class:`Lock` object is created and used as the underlying lock.
 
-   Locks also support the :ref:`context management protocol <async-with-locks>`.
+   Conditions also support the :ref:`context management protocol
```
Comment (1st1) [inline Doc/library/asyncio-sync.rst]: Last nit -- drop "also"

Comment (asvetlov) [reply Doc/library/asyncio-sync.rst]: Done

**Review by 1st1 — APPROVED** [review]
lgtm. Thanks!

**Pull request merged**